File: services/agent/src/agent/executor_client.py
"""REST client for the executor (the true backend).

Backend tools call the executor over REST — NEVER through the GraphQL facade.
GraphQL is the read/UI plane; writes triggered by the agent go straight to the
service that owns the datastore, through its compliance middleware.
"""

import contextlib
from typing import Any

import httpx

from .auth import AuthedUser

# Mirrors /contracts/fixtures/identity-headers.json (pinned by tests/test_contracts.py).
IDENTITY_HEADERS = {
    "callerService": "x-caller-service",
    "userId": "x-user-id",
    "userEmail": "x-user-email",
    "agentRunId": "x-agent-run-id",
    "toolName": "x-tool-name",
}

Json = dict[str, Any]


class ExecutorError(Exception):
    """status_code is set for HTTP-level failures (4xx/5xx), None for transport ones."""

    def __init__(self, message: str, status_code: int | None = None) -> None:
        super().__init__(message)
        self.status_code = status_code


class ExecutorClient:
    def __init__(self, base_url: str, client: httpx.AsyncClient | None = None) -> None:
        self._client = client or httpx.AsyncClient(base_url=base_url, timeout=10.0)

    async def aclose(self) -> None:
        await self._client.aclose()

    def _headers(self, user: AuthedUser, run_id: str, tool: str) -> dict[str, str]:
        headers = {
            IDENTITY_HEADERS["callerService"]: "agent",
            IDENTITY_HEADERS["userId"]: user.sub,
            IDENTITY_HEADERS["agentRunId"]: run_id,
            IDENTITY_HEADERS["toolName"]: tool,
        }
        if user.email:
            headers[IDENTITY_HEADERS["userEmail"]] = user.email
        return headers

    async def _call(
        self,
        user: AuthedUser,
        run_id: str,
        tool: str,
        method: str,
        path: str,
        body: Json | None = None,
    ) -> Any:
        try:
            res = await self._client.request(
                method, path, json=body, headers=self._headers(user, run_id, tool)
            )
        except httpx.HTTPError as exc:  # connection refused, timeout, ...
            raise ExecutorError(f"executor unreachable: {exc}") from exc
        if res.status_code >= 400:
            detail = res.text
            with contextlib.suppress(ValueError):
                detail = res.json().get("error", detail)
            raise ExecutorError(detail, status_code=res.status_code)
        return res.json()

    async def create_task(
        self, user: AuthedUser, run_id: str, title: str, due: str | None = None
    ) -> Json:
        body: Json = {"title": title}
        if due is not None:
            body["due"] = due
        return await self._call(user, run_id, "create_task", "POST", "/tasks", body)

    async def complete_task(self, user: AuthedUser, run_id: str, task_id: str) -> Json:
        return await self._call(user, run_id, "complete_task", "POST", f"/tasks/{task_id}/complete")

    async def list_tasks(self, user: AuthedUser, run_id: str) -> list[Json]:
        return await self._call(user, run_id, "list_tasks", "GET", "/tasks")

    async def rename_task(self, user: AuthedUser, run_id: str, task_id: str, title: str) -> Json:
        return await self._call(
            user, run_id, "rename_task", "POST", f"/tasks/{task_id}/rename", {"title": title}
        )

    async def set_due(self, user: AuthedUser, run_id: str, task_id: str, due: str | None) -> Json:
        return await self._call(
            user, run_id, "set_due", "POST", f"/tasks/{task_id}/due", {"due": due}
        )

    async def set_priority(
        self, user: AuthedUser, run_id: str, task_id: str, priority: str
    ) -> Json:
        return await self._call(
            user,
            run_id,
            "set_priority",
            "POST",
            f"/tasks/{task_id}/priority",
            {"priority": priority},
        )

    async def reopen_task(self, user: AuthedUser, run_id: str, task_id: str) -> Json:
        return await self._call(user, run_id, "reopen_task", "POST", f"/tasks/{task_id}/reopen")

    async def delete_task(self, user: AuthedUser, run_id: str, task_id: str) -> Json:
        return await self._call(user, run_id, "delete_task", "DELETE", f"/tasks/{task_id}")

    async def duplicate_task(self, user: AuthedUser, run_id: str, task_id: str) -> Json:
        return await self._call(
            user, run_id, "duplicate_task", "POST", f"/tasks/{task_id}/duplicate"
        )

    async def clear_completed(self, user: AuthedUser, run_id: str) -> Json:
        return await self._call(user, run_id, "clear_completed", "POST", "/tasks/completed/clear")

    async def create_tag(self, user: AuthedUser, run_id: str, name: str) -> Json:
        return await self._call(user, run_id, "create_tag", "POST", "/tags", {"name": name})

    async def tag_task(self, user: AuthedUser, run_id: str, task_id: str, name: str) -> Json:
        return await self._call(
            user, run_id, "tag_task", "POST", f"/tasks/{task_id}/tags", {"name": name}
        )

    async def reset_demo(self, user: AuthedUser, run_id: str) -> Json:
        return await self._call(user, run_id, "reset_demo", "POST", "/admin/reset")

File: services/graphql/src/executor-client.ts
import { IDENTITY_HEADERS } from '@mwe/contracts';
import type { AuthedUser } from './auth.js';

export interface TagDto {
  id: string;
  name: string;
  createdAt: string;
}

export interface TaskDto {
  id: string;
  title: string;
  due: string | null;
  completed: boolean;
  priority: 'LOW' | 'MEDIUM' | 'HIGH';
  tags: TagDto[];
  createdAt: string;
}

/**
 * The facade's ONLY data source is the executor's REST API — it has no
 * database access of its own. Each GraphQL operation maps to one allowlisted
 * executor tool name (see /contracts/identity-headers.md).
 */
export class ExecutorClient {
  constructor(private readonly baseUrl: string) {}

  private async call<T>(
    user: AuthedUser,
    toolName: string,
    method: 'GET' | 'POST' | 'DELETE',
    path: string,
    body?: unknown,
  ): Promise<T> {
    const headers: Record<string, string> = {
      [IDENTITY_HEADERS.callerService]: 'graphql',
      [IDENTITY_HEADERS.userId]: user.sub,
      [IDENTITY_HEADERS.toolName]: toolName,
    };
    if (user.email) headers[IDENTITY_HEADERS.userEmail] = user.email;
    if (body !== undefined) headers['content-type'] = 'application/json';

    const res = await fetch(`${this.baseUrl}${path}`, {
      method,
      headers,
      body: body === undefined ? undefined : JSON.stringify(body),
    });
    if (!res.ok) {
      const detail = await res.text();
      throw new Error(`executor ${method} ${path} failed with ${res.status}: ${detail}`);
    }
    return (await res.json()) as T;
  }

  private task(
    user: AuthedUser,
    tool: string,
    method: 'POST' | 'DELETE',
    path: string,
    body?: unknown,
  ) {
    return this.call<TaskDto>(user, tool, method, path, body);
  }

  tasks(user: AuthedUser): Promise<TaskDto[]> {
    return this.call(user, 'graphql.tasks', 'GET', '/tasks');
  }

  tags(user: AuthedUser): Promise<TagDto[]> {
    return this.call(user, 'graphql.tags', 'GET', '/tags');
  }

  createTask(user: AuthedUser, input: { title: string; due?: string | null }): Promise<TaskDto> {
    return this.task(user, 'graphql.createTask', 'POST', '/tasks', input);
  }

  completeTask(user: AuthedUser, id: string): Promise<TaskDto> {
    return this.task(
      user,
      'graphql.completeTask',
      'POST',
      `/tasks/${encodeURIComponent(id)}/complete`,
    );
  }

  renameTask(user: AuthedUser, id: string, title: string): Promise<TaskDto> {
    return this.task(
      user,
      'graphql.renameTask',
      'POST',
      `/tasks/${encodeURIComponent(id)}/rename`,
      {
        title,
      },
    );
  }

  setDue(user: AuthedUser, id: string, due: string | null): Promise<TaskDto> {
    return this.task(user, 'graphql.setDue', 'POST', `/tasks/${encodeURIComponent(id)}/due`, {
      due,
    });
  }

  setPriority(user: AuthedUser, id: string, priority: string): Promise<TaskDto> {
    return this.task(
      user,
      'graphql.setPriority',
      'POST',
      `/tasks/${encodeURIComponent(id)}/priority`,
      {
        priority,
      },
    );
  }

  reopenTask(user: AuthedUser, id: string): Promise<TaskDto> {
    return this.task(user, 'graphql.reopenTask', 'POST', `/tasks/${encodeURIComponent(id)}/reopen`);
  }

  deleteTask(user: AuthedUser, id: string): Promise<TaskDto> {
    return this.task(user, 'graphql.deleteTask', 'DELETE', `/tasks/${encodeURIComponent(id)}`);
  }

  duplicateTask(user: AuthedUser, id: string): Promise<TaskDto> {
    return this.task(
      user,
      'graphql.duplicateTask',
      'POST',
      `/tasks/${encodeURIComponent(id)}/duplicate`,
    );
  }

  clearCompleted(user: AuthedUser): Promise<{ deleted: TaskDto[] }> {
    return this.call(user, 'graphql.clearCompleted', 'POST', '/tasks/completed/clear');
  }

  createTag(user: AuthedUser, name: string): Promise<TagDto> {
    return this.call(user, 'graphql.createTag', 'POST', '/tags', { name });
  }

  tagTask(
    user: AuthedUser,
    id: string,
    name: string,
  ): Promise<{ task: TaskDto; tag: TagDto; tagCreated: boolean }> {
    return this.call(user, 'graphql.tagTask', 'POST', `/tasks/${encodeURIComponent(id)}/tags`, {
      name,
    });
  }

  resetDemo(user: AuthedUser): Promise<{ deletedTaskIds: string[]; deletedTagIds: string[] }> {
    return this.call(user, 'graphql.resetDemo', 'POST', '/admin/reset');
  }
}
